cluster-api/api/cluster/v1alpha1/types.go

```diff
@@ -301,3 +301,66 @@ type MachineList struct {
 	metav1.ListMeta `json:"metadata"`
 	Items           []Machine `json:"items"`
 }
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type MachineSet struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	Spec   MachineSetSpec   `json:"spec"`
+	Status MachineSetStatus `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type MachineSetList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []MachineSet `json:"items"`
+}
+
+// MachineSetSpec is the specification of a MachineSet.
+type MachineSetSpec struct {
+	// Replicas is the number of desired replicas.
+	// This is a pointer to distinguish between explicit zero and unspecified.
+	// Defaults to 1.
+	// +optional
+	Replicas *int32 `json:"replicas,omitempty"`
+
+	// Selector is a label query over machines that should match the replica count.
+	// Label keys and values that must match in order to be controlled by this MachineSet.
+	// It must match the machine template's labels.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
+	Selector *metav1.LabelSelector `json:"selector"`
```

Comment:
This should be required value, right? Auto-generated Selector doesn't work very well with `kubectl apply`

Also keeping it immutable can save us a lot of pain with orphaned resources.

Reply:
Yes, the Selector is a required value. 
Should i state that in the comment? By looking at the other fields it seems that everything is required which is not explicitly marked with `+optional`

Regarding immutability:
Based on the discussion in https://github.com/kubernetes/kubernetes/issues/50808
Yes, the selector should be immutable. 
Should this be stated in the comment? 
It seems its not common to state that in the comment. Other types just fail when trying to update.

Reply:
The immutability is enforced in the API server, so I guess comment is optional.

According to the [API Conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#optional-vs-required) `optional` fields should be `pointers` and marked with `// +optional`. `required` values should not be `pointers` and not be marked with `// +optional`

Reply:
changed it in d48e775b48bc772452058e30117e71c74e43106d